lib/pcr.c

```diff
@@ -161,3 +163,225 @@ TPM_RC get_max_supported_pcrs(TSS2_SYS_CONTEXT *sapi_context, UINT32 *max_pcrs)
+            LOG_ERR("Cannot open pcr-input-file %s", raw_pcrs_file);
+            return TPM_RC_NO_RESULT;
+        }
+       // Bank hashAlg values dictates the order of the list of digests
+        unsigned i;
+        for(i=0; i<pcr_values.count; i++) {
+            size_t sz = fread(&pcr_values.digests[i].t.buffer, 1, pcr_values.digests[i].t.size, fp);
+            if (sz != pcr_values.digests[i].t.size) {
+                const char *msg = ferror(fp) ? strerror(errno) :
+                        "end of file reached";
+                LOG_ERR("Reading from file \"%s\" failed: %s",
+                        raw_pcrs_file, msg);
+                fclose(fp);
+                return TPM_RC_NO_RESULT;
+            }
+        }
+        fclose(fp);
+    }
+
+    //If PCR input for policy is to be read from the TPM
+    if (!raw_pcrs_file) {
+        UINT32 pcr_update_counter;
+        TPML_PCR_SELECTION pcr_selection_out;
+        // Read PCRs
+        rval = Tss2_Sys_PCR_Read(sapi_context, 0,
+            &pcr_selections,
+            &pcr_update_counter, &pcr_selection_out, &pcr_values, 0);
+        if (rval != TPM_RC_SUCCESS) {
+            return rval;
+        }
+    }
+
+    // Calculate hashes
+    TPM2B_DIGEST pcr_digest =  TPM2B_TYPE_INIT(TPM2B_DIGEST, buffer);
+    rval = tpm_hash_sequence(sapi_context,
+        policy_session->authHash, pcr_values.count,
+        &pcr_values.digests[0], &pcr_digest);
+    if (rval != TPM_RC_SUCCESS) {
+        return rval;
+    }
+
+    // Call the PolicyPCR command
+    return Tss2_Sys_PolicyPCR(sapi_context, policy_session->sessionHandle,
+                              0, &pcr_digest, &pcr_selections, 0);
+}
+
+static TPM_RC start_policy_session (TSS2_SYS_CONTEXT *sapi_context,
+                                    SESSION **policy_session,
+                                    TPM_SE policy_session_type,
+                                    TPMI_ALG_HASH policy_digest_hash_alg) {
+    TPM2B_ENCRYPTED_SECRET encryptedSalt = TPM2B_EMPTY_INIT;
+    TPMT_SYM_DEF symmetric = {
+        .algorithm = TPM_ALG_NULL,
+    };
+    TPM2B_NONCE nonceCaller = TPM2B_EMPTY_INIT;
+    // Start policy session.
+    TPM_RC rval = tpm_session_start_auth_with_params(sapi_context,
+                                                     policy_session,
+                                                     TPM_RH_NULL, 0,
+                                                     TPM_RH_NULL, 0,
+                                                     &nonceCaller,
+                                                     &encryptedSalt,
+                                                     policy_session_type,
+                                                     &symmetric,
+                                                     policy_digest_hash_alg);
+    if (rval != TPM_RC_SUCCESS) {
+        LOG_ERR("Failed tpm session start auth with params\n");
+    }
+    return rval;
+}
+
+TPM_RC build_policy(TSS2_SYS_CONTEXT *sapi_context,
```

Comment:
@williamcroberts i was suggesting it to be more like tpm2_policy_pcr by moving the function to tpm2_policy(new file)